common/recursion_tracking.go

```diff
@@ -0,0 +1,33 @@
+package common
+
+import (
+	"reflect"
+	"strings"
+)
+
+const DefaultMaxRecursionDepth = 2
```

Comment:
For now we'll keep a global shared default max, in the future if necessary we can add a field in `ResourceProvider` such that a specific resource can have custom limit